# *PISA 2012 Data Wrangling*

In [201]:
# importing libraries
import numpy as np
import pandas as pd

## *Gather Data*

In [202]:
# importing dataset.
df = pd.read_csv('pisa2012.csv', encoding = 'cp1252')
df.head()

C:\Users\Shubham\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3063: DtypeWarning: Columns (15,16,17,21,22,23,24,25,26,30,31,36,37,45,65,123,155,156,157,158,159,160,161,162,163,164,165,166,167,168,169,170,171,284,285,286,287,288,289,290,291,292,293,294,295,296,297,298,299,300,301,302,303,307,308,309,310,311,312,313,314,315,316,317,318,319,320,321,322,323,324,325,326,327,328,329,330,331,332,333,334,335,336,337,338,339,340,341,342,343,344,345,346,347,348,349,350,351,352,353,354,355,356,357,376,377,378,379,380,381,382,383,384,385,386,387,388,389,390,391,392,393,394,395,396,397,398,399,400,401,402,403,475) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


Unnamed: 0      CNT  SUBNATIO  STRATUM      OECD       NC  SCHOOLID  \
0           1  Albania     80000  ALB0006  Non-OECD  Albania         1   
1           2  Albania     80000  ALB0006  Non-OECD  Albania         1   
2           3  Albania     80000  ALB0006  Non-OECD  Albania         1   
3           4  Albania     80000  ALB0006  Non-OECD  Albania         1   
4           5  Albania     80000  ALB0006  Non-OECD  Albania         1   

   STIDSTD  ST01Q01  ST02Q01  ...  W_FSTR75  W_FSTR76 W_FSTR77 W_FSTR78  \
0        1       10      1.0  ...   13.7954   13.9235  13.1249  13.1249   
1        2       10      1.0  ...   13.7954   13.9235  13.1249  13.1249   
2        3        9      1.0  ...   12.7307   12.7307  12.7307  12.7307   
3        4        9      1.0  ...   12.7307   12.7307  12.7307  12.7307   
4        5        9      1.0  ...   12.7307   12.7307  12.7307  12.7307   

   W_FSTR79 W_FSTR80 WVARSTRR VAR_UNIT SENWGT_STU  VER_STU  
0    4.3389  13.0829       19        1     0.2098  22NOV13  
1    4.3389  13.0829       19        1     0.2098  22NOV13  
2    4.2436  12.7307       19        1     0.1999  22NOV13  
3    4.2436  12.7307       19        1     0.1999  22NOV13  
4    4.2436  12.7307       19        1     0.1999  22NOV13  

[5 rows x 636 columns]

## Access Data

In [203]:
# Shape of the data
df.shape
# too large

(485490, 636)

In [204]:
# Info on dataset
df.info()
# Have to reduce this dataset

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 485490 entries, 0 to 485489
Columns: 636 entries, Unnamed: 0 to VER_STU
dtypes: float64(250), int64(18), object(368)
memory usage: 2.3+ GB


This dataset contains information from 485,490 students and 646 features.

***Due to having such large amount of features, I limited the number of features.***

*For my analysis, I wanted to focus on how gender, wealth and other factors plays a role in not only students' mathematics scores but also students' attitude towards mathematics. I also wanted to take a look at if the parents attitude towards children's mathematics differ among the gender. For these reasons, I have decided to focus on the following features:*

**Student Information**
+ CNT: Country
+ ST04Q01: Gender
+ ST13Q01:	Mother Highest Schooling
+ ST15Q01:	Mother Current Job Status
+ ST17Q01:	Father Highest Schooling
+ ST19Q01:	Father Current Job Status
+ COBN_F:	Country of Birth National Categories- Father
+ COBN_M:	Country of Birth National Categories- Mother
+ COBN_S:	Country of Birth National Categories- Self 
+ AGE:	Age of student   
+ EC05Q01:	First language learned
+ EC07Q01:	Language spoken - Mother
+ EC07Q02:	Language spoken - Father
+ EC07Q03:	Language spoken - Siblings
+ EC07Q04:	Language spoken - Best friend
+ EC07Q05:	Language spoken - Schoolmates  
+ WEALTH : Wealth
+ IMMIG:	Immigration status 

**Student Mathematics Scores**
+ PV1MATH:	Plausible value 1 in mathematics
+ PV2MATH:	Plausible value 2 in mathematics
+ PV3MATH:	Plausible value 3 in mathematics
+ PV4MATH:	Plausible value 4 in mathematics
+ PV5MATH:	Plausible value 5 in mathematics


**Student's Attitude Towards Mathematics**
+ **Instrumental Motivation**
    + ST29Q02:	Worthwhile for Work
    + ST29Q05:	Worthwhile for Career Chances
    + ST29Q07:	Important for Future Study
    + ST29Q08:	Helps to Get a Job


+ **Math Anxiety**
    + ST42Q01:	Worry That It Will Be Difficult
    + ST42Q03:	Get Very Tense
    + ST42Q05:	Get Very Nervous
    + ST42Q08:	Feel Helpless
    + ST42Q10:	Worry About Getting Poor <Grades>
	
+ **Math Self-Concept**
    + ST42Q02:	Not Good at Maths
    + ST42Q04:	Get Good (Grades)
    + ST42Q06:	Learn Quickly
    + ST42Q07:	One of Best Subjects
    + ST42Q09:	Understand Difficult Work

+ **Math Interest**
    + ST29Q01:	Enjoy Reading
    + ST29Q03:	Look Forward to Lessons
    + ST29Q04:	Enjoy Maths
    + ST29Q06:	Interested
    
+ **Math Work Ethic**	
    + ST46Q01:	Homework Completed in Time
    + ST46Q02:	Work Hard on Homework
    + ST46Q03:	Prepared for Exams
    + ST46Q04:	Study Hard for Quizzes
    + ST46Q05:	Study Until I Understand Everything
    + ST46Q06:	Pay Attention in Classes
    + ST46Q07:	Listen in Classes
    + ST46Q08:	Avoid Distractions When Studying
    + ST46Q09:	Keep Work Organized
    
+ **Math Behaviour**	
    + ST49Q01:	Talk about Maths with Friends
    + ST49Q02:	Help Friends with Maths
    + ST49Q03:	(Extracurricular) Activity
    + ST49Q04:	Participate in Competitions
    + ST49Q05:	Study More Than 2 Extra Hours a Day
    + ST49Q06:	Play Chess
    + ST49Q07:	Computer programming
    + ST49Q09:	Participate in Math Club


**Subjective Norms**
   + ST35Q04:	Parents Believe Studying Mathematics Is Important
   + ST35Q05:	Parents Believe Mathematics Is Important for Career
   + ST35Q06:	Parents Like Mathematics

**Reading Score**
+ PV1READ:	Plausible value 1 in reading
+ PV2READ:	Plausible value 2 in reading
+ PV3READ:	Plausible value 3 in reading
+ PV4READ:	Plausible value 4 in reading
+ PV5READ:	Plausible value 5 in reading
    
**Science Score**
+ PV1SCIE:	Plausible value 1 in science
+ PV2SCIE:	Plausible value 2 in science
+ PV3SCIE:	Plausible value 3 in science
+ PV4SCIE:	Plausible value 4 in science
+ PV5SCIE:	Plausible value 5 in science

In [205]:
# Collecting only columns which are required to save memory
df = df[['CNT', 'ST04Q01', 'ST13Q01', 'ST15Q01', 'ST17Q01', 'ST19Q01', 'COBN_F',
        'COBN_M', 'COBN_S', 'AGE', 'EC05Q01', 'EC07Q01', 'EC07Q02', 'EC07Q03', 'EC07Q04', 'EC07Q05', 'WEALTH', 'IMMIG',
        'PV1MATH','PV2MATH','PV3MATH','PV4MATH','PV5MATH', 'ST29Q02', 'ST29Q05', 'ST29Q07', 'ST29Q08', 'ST42Q01', 'ST42Q03'
        , 'ST42Q05', 'ST42Q08', 'ST42Q10', 'ST42Q02', 'ST42Q04', 'ST42Q06', 'ST42Q07', 'ST42Q09', 'ST29Q01', 'ST29Q03', 'ST29Q04', 'ST29Q06',
        'ST46Q01','ST46Q02','ST46Q03','ST46Q04','ST46Q05','ST46Q06','ST46Q07','ST46Q08','ST46Q09', 'ST49Q01', 'ST49Q02'
        , 'ST49Q03', 'ST49Q04', 'ST49Q05', 'ST49Q06', 'ST49Q07', 'ST49Q09', 'ST35Q04', 'ST35Q05', 'ST35Q06', 
        'PV1READ', 'PV2READ','PV3READ','PV4READ','PV5READ', 'PV1SCIE', 'PV2SCIE','PV3SCIE','PV4SCIE','PV5SCIE']]

In [206]:
# Checking for duplicates
df.duplicated().any()
# No duplicates

False

In [207]:
df['CNT'].value_counts()

Mexico                      33806
Italy                       31073
Spain                       25313
Canada                      21544
Brazil                      19204
                            ...  
Florida (USA)                1896
Perm(Russian Federation)     1761
Massachusetts (USA)          1723
Connecticut (USA)            1697
Liechtenstein                 293
Name: CNT, Length: 68, dtype: int64

**Looks like some are incorrectly labeled as a country when is a state or city (ie. Floria (USA))**

- **According to the survey only 65 countries participated.**
- ***Since our data is indicting 68 unique values under countries, this means that 3 values are not correct. Good thing we can easily see (at the bottom of our value count list) which 3 values are incorrectly labels! They are 'Florida (USA)', 'Massachusetts (USA)', 'Connecticut (USA)'.***

In [208]:
# These values to be removed and applied inside USA
df['CNT'].value_counts()[['United States of America','Florida (USA)','Massachusetts (USA)','Connecticut (USA)']].sum()

10294

***There should be a total of 10,294 students under 'United States of America'***

In [209]:
df['ST04Q01'].value_counts()

Female    245064
Male      240426
Name: ST04Q01, dtype: int64

***Gender is evenly split no missing gender data.***

In [210]:
# Checking Info
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 485490 entries, 0 to 485489
Data columns (total 71 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   CNT      485490 non-null  object 
 1   ST04Q01  485490 non-null  object 
 2   ST13Q01  457979 non-null  object 
 3   ST15Q01  467751 non-null  object 
 4   ST17Q01  443261 non-null  object 
 5   ST19Q01  451410 non-null  object 
 6   COBN_F   481825 non-null  object 
 7   COBN_M   481843 non-null  object 
 8   COBN_S   481836 non-null  object 
 9   AGE      485374 non-null  float64
 10  EC05Q01  129658 non-null  object 
 11  EC07Q01  44012 non-null   object 
 12  EC07Q02  43219 non-null   object 
 13  EC07Q03  42277 non-null   object 
 14  EC07Q04  42832 non-null   object 
 15  EC07Q05  42864 non-null   object 
 16  WEALTH   479597 non-null  float64
 17  IMMIG    471793 non-null  object 
 18  PV1MATH  485490 non-null  float64
 19  PV2MATH  485490 non-null  float64
 20  PV3MATH  485490 non-null  

**We have some missing values here**

In [211]:
# Too many column with null values
df.isnull().any().sum()

54

In [212]:
df.sample(10)

CNT ST04Q01                ST13Q01  \
442647            Slovenia    Male      <ISCED level 3A>    
354324            Malaysia    Male                    NaN   
421637  Russian Federation  Female  <ISCED level 3B, 3C>    
319093              Mexico    Male      <ISCED level 3A>    
30353            Australia    Male       <ISCED level 2>    
106034         Switzerland    Male      <ISCED level 3A>    
391322               Qatar  Female  <ISCED level 3B, 3C>    
410427   Connecticut (USA)    Male      <ISCED level 3A>    
58130               Brazil    Male       <ISCED level 1>    
448925      Chinese Taipei    Male  <ISCED level 3B, 3C>    

                                   ST15Q01                ST17Q01  \
442647        Working full-time <for pay>                     NaN   
354324        Working full-time <for pay>       <ISCED level 3A>    
421637  Other (e.g. home duties, retired)       <ISCED level 3A>    
319093         Working part-time <for pay>  <ISCED level 3B, 3C>    
30353          Working part-time <for pay>  <ISCED level 3B, 3C>    
106034         Working part-time <for pay>  <ISCED level 3B, 3C>    
391322  Other (e.g. home duties, retired)       <ISCED level 3A>    
410427        Working full-time <for pay>       <ISCED level 3A>    
58130         Working full-time <for pay>        <ISCED level 2>    
448925        Working full-time <for pay>        <ISCED level 2>    

                                   ST19Q01  \
442647                                 NaN   
354324         Working part-time <for pay>   
421637        Working full-time <for pay>    
319093  Not working, but looking for a job   
30353         Working full-time <for pay>    
106034         Working part-time <for pay>   
391322        Working full-time <for pay>    
410427        Working full-time <for pay>    
58130         Working full-time <for pay>    
448925        Working full-time <for pay>    

                                       COBN_F  \
442647                              Slovenia    
354324                              Malaysia    
421637  In another former USSR republic (RUS)   
319093                                Mexico    
30353                               Australia   
106034  One of the former Yugoslav republics    
391322                  Another country (QAT)   
410427              United States of America    
58130                                 Brazil    
448925                        Chinese Taipei    

                                       COBN_M  \
442647                              Slovenia    
354324                              Malaysia    
421637  In another former USSR republic (RUS)   
319093                                Mexico    
30353                               Australia   
106034  One of the former Yugoslav republics    
391322                  Another country (QAT)   
410427              United States of America    
58130                                 Brazil    
448925                        Chinese Taipei    

                                       COBN_S    AGE  ...   PV1READ   PV2READ  \
442647                              Slovenia   15.67  ...  229.9796  256.4438   
354324                              Malaysia   16.08  ...  271.0392  227.7341   
421637                    Russian Federation   16.25  ...  563.8847  551.9700   
319093                                Mexico   15.50  ...  530.0674  477.1391   
30353                               Australia  15.33  ...  580.0286  567.9994   
106034  One of the former Yugoslav republics   16.25  ...  376.0940  439.4476   
391322                  Another country (QAT)  15.58  ...  499.5456  583.7425   
410427              United States of America   15.67  ...  642.5803  711.5476   
58130                                 Brazil   16.08  ...  314.9858  322.2033   
448925                        Chinese Taipei   15.25  ...  553.5644  551.1586   

         PV3READ   PV4READ   PV5READ   PV1SCIE   PV2SCIE   PV3SCIE   PV4SCIE  \
442647  262.0574  249.2263  236.

### Access Data Summary

+ Delete unneccessary features
+ Rename feature labels
+ Correctly categorize countries under CNT (country): Replace values of 'Florida (USA)','Massachusetts (USA)', and 'Connecticut (USA)' with 'United States of America'
+ Student and Parent's Attitudes: Combine the features in each section into one feature by assigning point values and taking average.
  + If student did not answer at least one feature under each section, then fill in Null Value for section's overall score
  + Becareful when assigning point values due to fact that "Not Good at Math" is sometime negative while the remainder are positive. In this case, for 'Not good at Math' (ST42Q02) we would assign 'Strongly Agree' a point value of 1, 'Agree' a point value of 2 and so on.
+ Taking average of all the plausible scores.

## Clean Data

**Delete unneccessary features by making copy of dataframe with neccessary features**

In [213]:
# Creating new dataframe
df_clean = df.copy()
# Deleting the old dataframe
del df

In [214]:
df_clean.sample(10)

CNT ST04Q01                                ST13Q01  \
19198        Argentina    Male                      <ISCED level 3A>    
322604          Mexico  Female  She did not complete <ISCED level 1>    
246170           Italy  Female                      <ISCED level 3A>    
245683           Italy    Male                      <ISCED level 3A>    
131516  Czech Republic    Male                      <ISCED level 3A>    
241430           Italy    Male                      <ISCED level 3A>    
156724           Spain  Female                  <ISCED level 3B, 3C>    
327668          Mexico    Male                       <ISCED level 1>    
454068  Chinese Taipei  Female                      <ISCED level 3A>    
224942       Indonesia    Male                      <ISCED level 3A>    

                                   ST15Q01  \
19198                                  NaN   
322604         Working part-time <for pay>   
246170        Working full-time <for pay>    
245683        Working full-time <for pay>    
131516  Other (e.g. home duties, retired)    
241430        Working full-time <for pay>    
156724         Working part-time <for pay>   
327668  Other (e.g. home duties, retired)    
454068        Working full-time <for pay>    
224942  Other (e.g. home duties, retired)    

                                     ST17Q01                       ST19Q01  \
19198                                    NaN   Working part-time <for pay>   
322604  He did not complete <ISCED level 1>   Working full-time <for pay>    
246170                      <ISCED level 2>   Working full-time <for pay>    
245683                      <ISCED level 2>   Working full-time <for pay>    
131516                 <ISCED level 3B, 3C>    Working part-time <for pay>   
241430                     <ISCED level 3A>   Working full-time <for pay>    
156724                 <ISCED level 3B, 3C>   Working full-time <for pay>    
327668                     <ISCED level 3A>   Working full-time <for pay>    
454068                      <ISCED level 2>   Working full-time <for pay>    
224942                      <ISCED level 2>                            NaN   

                 COBN_F           COBN_M           COBN_S    AGE  ...  \
19198         Argentina        Argentina        Argentina  15.92  ...   
322604          Mexico           Mexico           Mexico   15.33  ...   
246170            Italy            Italy            Italy  15.67  ...   
245683            Italy            Italy            Italy  16.00  ...   
131516  Czech Republic   Czech Republic   Czech Republic   15.50  ...   
241430            Italy            Italy            Italy  15.58  ...   
156724  La Rioja (ESP)   La Rioja (ESP)   La Rioja (ESP)   15.42  ...   
327668          Mexico           Mexico           Mexico   15.75  ...   
454068  Chinese Taipei   Chinese Taipei   Chinese Taipei   15.50  ...   
224942        Indonesia        Indonesia        Indonesia  16.25  ...   

         PV1READ   PV2READ   PV3READ   PV4READ   PV5READ   PV1SCIE   PV2SCIE  \
19198   335.8363  398.3881  364.7064  246.8204  294.9371  387.6725  355.9679   
322604  435.4448  447.3594  471.1887  455.3025  426.7073  434.2969  399.7949   
246170  529.1733  574.4490  529.9677  477.5432  511.6985  430.5670  548.9929   
245683  420.7623  446.4246  487.3238  441.6129  462.4635  445.6733  514.6774   
131516  473.9313  416.1912  427.4185  441.0515  445.8632  523.2562  493.4166   
241430  573.6131  596.0675  585.6422  595.2656  591.2559  577.9000  606.8072   
156724  387.7861  491.8408  375.0772  459.2741  424.3244  493.3234  490.5259   
327668  424.3711  424.3711  486.9228  430.7866  515.7928  515.4233  485.5837   
454068  601.3761  577.5468  571.9867  601.3761  566.4265  529.4107  536.8706   
224942  408.7331  403.1195  357.4087  423.1682  404.7234  429.8210  376.6692   

         PV3SCIE   PV4SCIE   PV5SCIE  
19198   353.1705  263.6516  249.6643  
322604  438.0269  410.9847  396.0649  
246170  480.9213  466.9340  455.7441  
245683  

**Combine the features in each section into one feature by assigning point values and taking average. This will create a score which will be average of all type of attributes.**

(If student did not answer at least one feature under each section, then fill in Null Value for section's overall score.)

**Sections**
+ **Instrumental Motivation** 
     + worthwhile for work, worthwhile for career changes, important for future study
        
        
+ **Math Anxiety** 
     + worry that it will be difficult, get very tense, get very nervous, feel helpless, worry about getting poor grades
        
        
+ **Math Self-Concept** 
     + not good at Math, get good grades, learn quickly, one of best subjects, understand difficult work
      
      
+ **Math Interest** 
     + enjoy reading, look forward to lessons, enjoy Math, interested in Math
        
        
+ **Math Work Ethic** 
     + homework completed on time, work hard on homework, prepared for exams, study hard for quizzes, study until I understand everything, pay attention in classes, listen in classes, avoid distractions when studying, keep work organized
        
        
+ **Math Behaviour** 
     + talk about Math with friends, help friends with Math, Extracurricular Activity, participate in competitions, study more than 2 extra hours a day, play chess, computer programming, participate in math club


+ **Student's Parental View of Mathematics**
    + parents believe studying mathematics is important, parents believe mathematics is important for career, parents like mathematics


In [215]:
# Worthwhile for Work, ST29Q05: Worthwhile for Career Chances, Important for Future Study
# Helps to Get a Job
motivation = ['ST29Q02','ST29Q05','ST29Q07','ST29Q08']

# Worry That It Will Be Difficult, Get Very Tense, Get Very Nervous
# Feel Helpless, Worry About Getting Poor <Grades>
anxiety = ['ST42Q01','ST42Q03','ST42Q05','ST42Q08','ST42Q10']

# Not Good at Maths, Get Good (Grades), Learn Quickly
# One of Best Subjects, Understand Difficult Work
self = ['ST42Q02','ST42Q04','ST42Q06','ST42Q07','ST42Q09']

# Enjoy Reading, Look Forward to Lessons, Enjoy Maths, Interested
interest = ['ST29Q01','ST29Q03','ST29Q04','ST29Q06']

# Homework Completed in Time, Work Hard on Homework, Prepared for Exams
# Study Hard for Quizzes, Study Until I Understand Everything, Pay Attention in Classes
# Listen in Classes, Avoid Distractions When Studying, Keep Work Organized
work_ethic = ['ST46Q01','ST46Q02','ST46Q03','ST46Q04','ST46Q05','ST46Q06','ST46Q07','ST46Q08','ST46Q09']

# alk about Maths with Friends, Help Friends with Maths, (Extracurricular) Activity
# Participate in Competitions, Study More Than 2 Extra Hours a Day, Play Chess
# Computer programming, Participate in Math Club
behavior = ['ST49Q01','ST49Q02','ST49Q03','ST49Q04','ST49Q05','ST49Q06','ST49Q07','ST49Q09']

# Parents Believe Studying Mathematics Is Important
# Parents Believe Mathematics Is Important for Career
# Parents Like Mathematics
parents = ['ST35Q04','ST35Q05','ST35Q06']

***Instrumental Motivation, Math Anxiety, Math Interest, Math Work Ethic, Student's Parental Attitude of Mathematics will be using the same scale where 'Strongly agree' would be 4 points and 'Strongly disagree' would be 1 point***



In [216]:
# Sections Instrumental Motivation, Math Anxiety, Math Interest, and Math Work Ethic
# will be using the same scale
points = {'Strongly agree':4,'Agree': 3,'Disagree': 2, 'Strongly disagree':1}

In [217]:
# Function that will assign point values to each feature category
# and create/combine into one feature
# features variabe will be a list argument
# cmb_name variable will be the name of the new combined feature (string form)
# scale variable is what dictionary point system we will be using

def combine_features(features, cmb_name, scale):
    for feature in features:
        df_clean.loc[:, feature] = df_clean.loc[:, feature].map(scale)
    df_clean[cmb_name] = df_clean.loc[:, features].sum(axis = 1, skipna = False)/len(features)

In [218]:
# Instrumental Motivation
combine_features(motivation,'motivation',points)

# Math Anxiety
combine_features(anxiety, 'anxiety', points)

# Math Interest
combine_features(interest, 'interest', points)

# Math Work Ethic
combine_features(work_ethic, 'work_ethic', points)

# Parents
combine_features(parents, 'parents', points)

In [219]:
df_clean[['motivation','anxiety','interest','work_ethic', 'parents']].head(10)

motivation  anxiety  interest  work_ethic   parents
0        3.50      2.6      3.00    3.000000  3.000000
1        2.75      NaN      2.25    3.222222  2.666667
2        4.00      NaN      3.25    3.888889  3.666667
3         NaN      NaN       NaN         NaN       NaN
4        4.00      3.2      2.50    3.777778  3.666667
5         NaN      2.2       NaN         NaN       NaN
6         NaN      2.8       NaN         NaN       NaN
7        3.50      2.4      3.00    3.666667  3.333333
8        3.00      2.4      3.00    3.000000  3.333333
9         NaN      1.4       NaN         NaN       NaN

***Math Behaviour would have a different points system due to being categorized differently. Features under this section contains categories Always or Almost Always, Often, Sometimes, Never or Rarely. In this scenario, we would have 'Always or Almost Always' assigned to 4 points and 'Often' to 3 points, and so on.***

In [220]:
# points system for Math Behavious
points = {'Always or almost always': 4, 'Often':3, 'Sometimes':2, 'Never or rarely':1}

In [221]:
combine_features(behavior,'behavior',points)

In [222]:
df_clean['behavior'].head()

0    1.75
1    2.50
2    2.00
3     NaN
4     NaN
Name: behavior, dtype: float64

In [223]:
df_clean[behavior].head()

ST49Q01  ST49Q02  ST49Q03  ST49Q04  ST49Q05  ST49Q06  ST49Q07  ST49Q09
0      3.0      2.0      2.0      2.0      2.0      1.0      1.0      1.0
1      2.0      3.0      4.0      2.0      4.0      1.0      1.0      3.0
2      2.0      4.0      2.0      1.0      4.0      1.0      1.0      1.0
3      NaN      NaN      NaN      NaN      NaN      NaN      NaN      NaN
4      4.0      4.0      3.0      3.0      2.0      NaN      2.0      2.0

***Section Math Self-Concept would be an exception from the above scale system due to one feature asks students if they believe they are not good at Math meaning 'Strongly agree' would be a negative view of themselves. The other features under Math Self Concept have 'Strongly agree' as a positive view of themselves. This would be we would reverse the point system for just the feature 'Not Good at Maths' (ST42Q02)***



In [224]:
# reverse point system for 'Not Good at Math' feature
df_clean.loc[:,'ST42Q02'] = df_clean.loc[:,'ST42Q02'].map({'Strongly agree':1, 
                                             'Agree': 2,'Disagree': 3, 'Strongly disagree':4})

In [225]:
# regular point system
for _ in ['ST42Q04','ST42Q06','ST42Q07','ST42Q09']:
    df_clean.loc[:,_] = df_clean.loc[:,_].map({'Strongly agree':4,'Agree': 3,'Disagree': 2, 'Strongly disagree':1})

In [226]:
df_clean[self].head()

ST42Q02  ST42Q04  ST42Q06  ST42Q07  ST42Q09
0      3.0      3.0      3.0      3.0      2.0
1      NaN      NaN      NaN      NaN      NaN
2      NaN      NaN      NaN      NaN      NaN
3      1.0      3.0      2.0      4.0      3.0
4      1.0      4.0      2.0      2.0      3.0

In [227]:
#create new feature that takes average of Self Features
df_clean['self'] = df_clean.loc[:, self].sum(axis = 1, skipna = False)/len(self)

In [228]:
df_clean['self'].head()

0    2.8
1    NaN
2    NaN
3    2.6
4    2.4
Name: self, dtype: float64

In [229]:
df_clean[self].head()

ST42Q02  ST42Q04  ST42Q06  ST42Q07  ST42Q09
0      3.0      3.0      3.0      3.0      2.0
1      NaN      NaN      NaN      NaN      NaN
2      NaN      NaN      NaN      NaN      NaN
3      1.0      3.0      2.0      4.0      3.0
4      1.0      4.0      2.0      2.0      3.0

**Calculating Average of all the plausible score.**

In [230]:
df_clean['avg_math_score'] = (df_clean['PV1MATH'] + df_clean['PV2MATH'] + df_clean['PV3MATH'] + df_clean['PV4MATH'] + df_clean['PV5MATH']) / 5
df_clean['avg_reading_score'] = (df_clean['PV1READ'] + df_clean['PV2READ'] + df_clean['PV3READ'] + df_clean['PV4READ'] + df_clean['PV5READ']) / 5
df_clean['avg_science_score'] = (df_clean['PV1SCIE'] + df_clean['PV2SCIE'] + df_clean['PV3SCIE'] + df_clean['PV4SCIE'] + df_clean['PV5SCIE']) / 5

**Delete unneccessary features**

In [231]:
df_clean.drop(motivation + anxiety + self + interest + work_ethic + behavior + parents, axis = 1,inplace = True)
df_clean.drop(columns = ['PV1MATH', 'PV2MATH', 'PV3MATH', 'PV4MATH', 'PV5MATH', 'PV1READ', 'PV2READ', 'PV3READ', 'PV4READ',
                 'PV5READ', 'PV1SCIE', 'PV2SCIE', 'PV3SCIE', 'PV4SCIE', 'PV5SCIE'], inplace = True)
df_clean.head()

CNT ST04Q01                                ST13Q01  \
0  Albania  Female                      <ISCED level 3A>    
1  Albania  Female                      <ISCED level 3A>    
2  Albania  Female                  <ISCED level 3B, 3C>    
3  Albania  Female                  <ISCED level 3B, 3C>    
4  Albania  Female  She did not complete <ISCED level 1>    

                              ST15Q01                ST17Q01  \
0  Other (e.g. home duties, retired)       <ISCED level 3A>    
1        Working full-time <for pay>       <ISCED level 3A>    
2        Working full-time <for pay>       <ISCED level 3A>    
3        Working full-time <for pay>       <ISCED level 3A>    
4         Working part-time <for pay>  <ISCED level 3B, 3C>    

                        ST19Q01   COBN_F   COBN_M   COBN_S    AGE  ...  \
0   Working part-time <for pay>  Albania  Albania  Albania  16.17  ...   
1  Working full-time <for pay>   Albania  Albania  Albania  16.17  ...   
2  Working full-time <for pay>   Albania  Albania  Albania  15.58  ...   
3  Working full-time <for pay>   Albania  Albania  Albania  15.67  ...   
4   Working part-time <for pay>  Albania  Albania  Albania  15.50  ...   

  motivation anxiety interest work_ethic   parents behavior  self  \
0       3.50     2.6     3.00   3.000000  3.000000     1.75   2.8   
1       2.75     NaN     2.25   3.222222  2.666667     2.50   NaN   
2       4.00     NaN     3.25   3.888889  3.666667     2.00   NaN   
3        NaN     NaN      NaN        NaN       NaN      NaN   2.6   
4       4.00     3.2     2.50   3.777778  3.666667      NaN   2.4   

  avg_math_score  avg_reading_score  avg_science_score  
0      366.18634          261.01424          371.91348  
1      470.56396          384.68832          478.12382  
2      505.53824          405.18154          486.60946  
3      449.45476          477.46376          453.97240  
4      385.50398          256.01010          367.15778  

[5 rows x 28 columns]

In [232]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 485490 entries, 0 to 485489
Data columns (total 28 columns):
 #   Column             Non-Null Count   Dtype  
---  ------             --------------   -----  
 0   CNT                485490 non-null  object 
 1   ST04Q01            485490 non-null  object 
 2   ST13Q01            457979 non-null  object 
 3   ST15Q01            467751 non-null  object 
 4   ST17Q01            443261 non-null  object 
 5   ST19Q01            451410 non-null  object 
 6   COBN_F             481825 non-null  object 
 7   COBN_M             481843 non-null  object 
 8   COBN_S             481836 non-null  object 
 9   AGE                485374 non-null  float64
 10  EC05Q01            129658 non-null  object 
 11  EC07Q01            44012 non-null   object 
 12  EC07Q02            43219 non-null   object 
 13  EC07Q03            42277 non-null   object 
 14  EC07Q04            42832 non-null   object 
 15  EC07Q05            42864 non-null   object 
 16  WE

**Rename feature labels**

In [233]:
col_names = {'CNT':'country', 'ST04Q01':'gender', 'ST13Q01':'mother_school', 'ST15Q01':'mother_job', 'ST17Q01':'father_school', 'ST19Q01':'father_job',
            'COBN_F':'father_nation', 'COBN_M':'mother_nation','COBN_S':'student_nation', 
            'EC05Q01':'first_lang_learn', 'EC07Q01':'mother_lang','EC07Q02':'father_lang','EC07Q03':'siblings_lang','EC07Q04':'friend_lang','EC07Q05':'school_lang',
            'AGE':'age', 'WEALTH':'wealth', 'IMMIG':'immigration'}

In [234]:
df_clean.rename(columns = col_names, inplace = True)

In [235]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 485490 entries, 0 to 485489
Data columns (total 28 columns):
 #   Column             Non-Null Count   Dtype  
---  ------             --------------   -----  
 0   country            485490 non-null  object 
 1   gender             485490 non-null  object 
 2   mother_school      457979 non-null  object 
 3   mother_job         467751 non-null  object 
 4   father_school      443261 non-null  object 
 5   father_job         451410 non-null  object 
 6   father_nation      481825 non-null  object 
 7   mother_nation      481843 non-null  object 
 8   student_nation     481836 non-null  object 
 9   age                485374 non-null  float64
 10  first_lang_learn   129658 non-null  object 
 11  mother_lang        44012 non-null   object 
 12  father_lang        43219 non-null   object 
 13  siblings_lang      42277 non-null   object 
 14  friend_lang        42832 non-null   object 
 15  school_lang        42864 non-null   object 
 16  we

***Correctly categorize countries country: Replace values of 'Florida (USA)','Massachusetts (USA)', and 'Connecticut (USA)' with 'United States of America'***

In [236]:
df_clean.country.value_counts()[['United States of America',
                                 'Florida (USA)','Massachusetts (USA)',
                                 'Connecticut (USA)' ]]

United States of America    4978
Florida (USA)               1896
Massachusetts (USA)         1723
Connecticut (USA)           1697
Name: country, dtype: int64

In [237]:
states = ['Florida (USA)','Massachusetts (USA)','Connecticut (USA)']

df_clean['country'].replace(states, 'United States of America', inplace = True)

In [238]:
# Checking if there are only 65 countries.
len(df_clean['country'].unique())

65

In [239]:
df_clean.head()

country  gender                          mother_school  \
0  Albania  Female                      <ISCED level 3A>    
1  Albania  Female                      <ISCED level 3A>    
2  Albania  Female                  <ISCED level 3B, 3C>    
3  Albania  Female                  <ISCED level 3B, 3C>    
4  Albania  Female  She did not complete <ISCED level 1>    

                           mother_job          father_school  \
0  Other (e.g. home duties, retired)       <ISCED level 3A>    
1        Working full-time <for pay>       <ISCED level 3A>    
2        Working full-time <for pay>       <ISCED level 3A>    
3        Working full-time <for pay>       <ISCED level 3A>    
4         Working part-time <for pay>  <ISCED level 3B, 3C>    

                     father_job father_nation mother_nation student_nation  \
0   Working part-time <for pay>       Albania       Albania        Albania   
1  Working full-time <for pay>        Albania       Albania        Albania   
2  Working full-time <for pay>        Albania       Albania        Albania   
3  Working full-time <for pay>        Albania       Albania        Albania   
4   Working part-time <for pay>       Albania       Albania        Albania   

     age  ... motivation anxiety interest work_ethic   parents behavior  self  \
0  16.17  ...       3.50     2.6     3.00   3.000000  3.000000     1.75   2.8   
1  16.17  ...       2.75     NaN     2.25   3.222222  2.666667     2.50   NaN   
2  15.58  ...       4.00     NaN     3.25   3.888889  3.666667     2.00   NaN   
3  15.67  ...        NaN     NaN      NaN        NaN       NaN      NaN   2.6   
4  15.50  ...       4.00     3.2     2.50   3.777778  3.666667      NaN   2.4   

  avg_math_score  avg_reading_score  avg_science_score  
0      366.18634          261.01424          371.91348  
1      470.56396          384.68832          478.12382  
2      505.53824          405.18154          486.60946  
3      449.45476          477.46376          453.97240  
4      385.50398          256.01010          367.15778  

[5 rows x 28 columns]

In [240]:
# Filling "Missing and None" if any with null values
df_clean.replace(to_replace = {"Missing":np.nan}, inplace = True)
df_clean.replace(to_replace = {"missing":np.nan}, inplace = True)
df_clean.replace(to_replace = {"None":np.nan}, inplace = True)

In [241]:
# changing age type to number
df_clean.age = df_clean.age.apply(np.ceil)
df_clean = df_clean.apply(pd.to_numeric, errors = 'ignore')

### Not removing null values because this might create problem with other variables.

### Exporting data to machine.

In [242]:
# Exporting clean dataset to machine
df_clean.to_csv('pisa2012_clean.csv', index = False)